arrows/vxl/convert_image.cxx

```diff
@@ -0,0 +1,508 @@
+  }
+  else
+  {
+    scale = static_cast< double >( max_val ) /
+            std::numeric_limits< InputType >::max();
+  }
+
+  unsigned ni = src.ni(), nj = src.nj(), np = src.nplanes();
+  dst.set_size( ni, nj, np );
+
+  std::ptrdiff_t istep = src.istep(), jstep = src.jstep(),
+                 pstep = src.planestep();
+
+  const InputType* plane = src.top_left_ptr();
+  for( unsigned p = 0; p < np; ++p, plane += pstep )
+  {
+    const InputType* row = plane;
+    for( unsigned j = 0; j < nj; ++j, row += jstep )
+    {
+      const InputType* pixel = row;
+      for( unsigned i = 0; i < ni; ++i, pixel += istep )
+      {
+        if( *pixel < lower_bound )
+        {
+          dst( i, j, p ) = 0;
+        }
+        else if( *pixel > upper_bound )
+        {
+          dst( i, j, p ) = std::numeric_limits< OutputType >::max();
+        }
+        else
+        {
+          dst( i, j,
+               p ) =
+            static_cast< OutputType >( ( *pixel - lower_bound ) *
+                                       scale );
+        }
+      }
+    }
+  }
+}
+
+} // end anonoymous namespace
+
+// ----------------------------------------------------------------------------
+/// Private implementation class
+class convert_image::priv
+{
+public:
+
+  priv()
+    : format( "byte" )
+      , single_channel( false )
+      , scale_factor( 0.0 )
+      , random_greyscale( 0.0 )
+      , percentile_norm( -1.0 )
+  {
+  }
+
+  ~priv()
+  {
+  }
+
+  std::string format;
+  bool single_channel;
+  double scale_factor;
+  double random_greyscale;
+  double percentile_norm;
+};
+
+// ----------------------------------------------------------------------------
+convert_image
+::convert_image()
+  : d( new priv() )
+{
+  attach_logger( "arrows.vxl.convert_image" );
+}
+
+convert_image
+::~convert_image()
+{
+}
+
+vital::config_block_sptr
+convert_image
+::get_configuration() const
+{
+  // get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value( "format", d->format,
+                     "Output type format: byte, sbyte, float, double, uint16, uint32, etc." );
+  config->set_value( "single_channel", d->single_channel,
+                     "Convert input (presumably multi-channel) to contain a single channel, using "
+                     "either standard RGB to grayscale conversion weights, or averaging." );
+  config->set_value( "scale_factor", d->scale_factor,
+                     "Optional input value scaling factor" );
+  config->set_value( "random_greyscale", d->random_greyscale,
+                     "Convert input image to a 3-channel greyscale image randomly with this percentage "
+                     "between 0.0 and 1.0. This is used for machine learning augmentation." );
+  config->set_value( "percentile_norm", d->percentile_norm,
+                     "If set, perform percentile normalization such that the output image's min and max "
```

Comment:
I don't think this is worded well. It sounds like this *reduces* the image contrast?

Reply:
This function took me longer than I'd have liked to interpret. It basically uses this provided `fraction`, which is between [`0.0`,` 0.5`), so right of the bat the name is missleading. Then it finds the quintiles at `fraction` and `1 - fraction`. Then the output image is scaled and clamped so that the lower quintile is scaled to the lower bound of the output and the upper quintile is scaled to the upper bound.

 I'm not sure if my updated description is effective at describing the behavior so I'd welcome other suggestions.